### PR TITLE
feat(common/models): initial integration for enhanced corrections

### PR DESCRIPTION
**Pull request by jahorton**
Provides an initial resolution of #2223.

Introduces a much stronger 'correction' algorithm for use in predictive text.  Additionally, the engine is now capable of correcting errors anywhere within a word, rather than only at its end.
- Note that further optimizations, pruning, and thresholding strategies will be needed to ensure proper responsiveness.
- There are also a lot of remaining 'rough edges'. 
   - By far the worst 'rough edge':  ensuring the user can actually _apply_ the suggestion.  (Addressed by #3556)
   - Second 'rough edge':  the current integration level is only guaranteed to work for the first word a user types.
      - It's actually working beyond that in current tests, but I haven't thoroughly vetted that behavior yet.
      - Now more thoroughly vetted (as of #3556)
   - Third - there's some serious 'aliasing' in the connection between the ContextTracker component and its SearchSpace field at present.

Tends to run faster when input is more accurate.  As a result, testing with the following input may take a bit of a wait at present:

Screen Shot 2020-09-03 at 2 25 23 PM

Naturally, it helps if you 'click' on the edges of the keys closest to the suggested characters, as it's closer to the 'intended' input.

----

Note that in its current form, this 'integration' fully replaces the old correction algorithm within KeymanWeb.  As such, this particular PR should not be merged until most/all of the 'rough edges' have been sanded off - especially regarding the context handling / "first word only" bit mentioned above.
- Fortunately, #3556 provides a very significant amount of said 'sanding'.  Feel free to test the two versions to see the difference; #3556 works better as a general-use "live demo" as a result of its tweaks.

There are also a lot of bug fixes for some of the as-of-yet unmerged parent PRs here... I might should port them back to their respective PRs as well, but as it'd be pretty involved, I'm not sure it's worth the additional effort.

### CODE REVIEW

**Comment by eddieantonio** [comment]
@jahorton, should I wait for the other PRs before reviewing this? (e.g., #3556, #3565)

**Comment by jahorton** [comment]
> 
> 
> @jahorton, should I wait for the other PRs before reviewing this? (e.g., #3556, #3565)

I figure that it should be easier to review with them separate, rather than merged.  At a minimum, the others should be simpler to follow; this one does most of the complex 'wiring' needed for integration.

**Comment by jahorton** [comment]
Hopefully you're ready to take another pass at this now, @eddieantonio, having reviewed some of the later PRs in the chain?  Can't merge them to `master` without this one merging first.

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 1]
```diff
@@ -63,12 +63,32 @@ namespace correction {
...
+      // That's a really, really high likelihood, thoough.
+      //
+      // At any rate, we can linearly scale the known-cost to have about whatever probability we want.
+      // If we can state it as p = 1 / (c * e), note then that ln(c * e) = ln(c) + 1.  So, scale * (ln(c) + 1).
+      // If we can state it as e^x, note that ln(e^x) = x * ln(e) = x - just scale by 'x'!
+
+      // p = 1 / (e^4) = 0.01831563888.  This still exceeds many neighboring keys!
+      // p = 1 / (e^5) = 0.00673794699.  Strikes a good balance.
+      // Should easily give priority to neighboring keys before edit-distance kicks in (when keys are a bit ambiguous)
+      return SearchSpace.EDIT_DISTANCE_COST_SCALE * this.knownCost + this.inputSamplingCost;
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:91]: Really good commenting!

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 2]
```diff
@@ -63,12 +63,32 @@ namespace correction {
       // TODO:  Optimize so that we're not frequently recomputing this?
       // TODO:  We might should generalize this so that the probability-to-cost function isn't directly hard-coded.
       //        Seems like a decent first conversion function though, at least.
-      return this.optimalInput.map(mass => mass.p).reduce((previous, current) => previous + (1 - current), 0);
+      //
+      // NOTE:  changes here need to be mirrored in the Node cost function, too!
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:67]: Is there a plan to reduce the duplication between, e.g., `inputSamplingCost()` and `currentCost()` in `SearchNode` or `SearchEdge`?

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 3]
```diff
@@ -237,6 +274,8 @@ namespace correction {
   // The set of search spaces corresponding to the same 'context' for search.
   // Whenever a wordbreak boundary is crossed, a new instance should be made.
   export class SearchSpace {
+    static readonly EDIT_DISTANCE_COST_SCALE = 5;
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:277]: I feel like the comments in `currentCost()` could be helpful here instead (or in addition!) — why was 5 chosen? Well, that's answered in `(SearchNode || SearchEdge).currentCost()`!

**Review by eddieantonio — APPROVED** [review]
LGTM! I have some concerns about some code duplication, but no concrete way to address it 😅

File: common/predictive-text/worker/correction/distance-modeler.ts (hunk 2 above)
Comment (jahorton) [reply common/predictive-text/worker/correction/distance-modeler.ts:67]: You already approved it.  https://github.com/keymanapp/keyman/pull/3565

File: common/predictive-text/worker/correction/distance-modeler.ts (hunk 3 above)
Comment (jahorton) [reply common/predictive-text/worker/correction/distance-modeler.ts:277]: Went ahead and edited that into #3565.

**Pull request merged**